### PR TITLE
Local path & source URL installation support

### PR DESCRIPTION
**Pull request by codecontribu**
Local path & Source URL installation support:

- Changes to pass local keycloak zip file.
-  Changes to pass local RHSSO zip file.
- Changes to pass source url to download RHSSO zip file from.

Now all following scenarios are available:
1. RHSSO from CSP
2. RHSSO from remote source
3. RHSSO from localpath
4. Keycloak from remote source
5. Keycloak from local path

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -31,11 +31,23 @@ collections:
 
 ### Install Playbook
 
-`playbooks/keycloak.yml` installs the keycloak or Red Hat Single Sign-On based on the defined variables.
+`playbooks/keycloak.yml` installs the keycloak or Red Hat Single Sign-On(RHSSO) based on the defined variables.
 
-### Choosing between Red Hat products and upstream project
+### Choosing between Red Hat products and upstream(Keycloak) project
```
Comment (sabre1041) [inline README.md]:
```suggestion
### Choosing between Red Hat products and upstream (Keycloak) project
```

File: README.md [hunk 2]
```diff
@@ -31,11 +31,23 @@ collections:
...
 
-The roles supports installing Red Hat Single Sign-On from the Customer Portal, when the following variables are defined:
+The roles supports installing Keycloak or Red Hat Single Sign-On in following ways
```
Comment (sabre1041) [inline README.md]:
```suggestion
The roles supports installing Keycloak or Red Hat Single Sign-On in the following ways
```

File: README.md [hunk 3]
```diff
@@ -31,11 +31,23 @@ collections:
...
+
+#### Install upstream from remote source
+
+This is default way, no need to define any additional variables.
+
+#### Install upstream from local source when the following variable is defined
+
+```
+keycloak_zip_file_local_path: <local path of keycloak zip file>
```
Comment (sabre1041) [inline README.md]: Need to denote that this is local on the Ansible control node

File: README.md [hunk 4]
```diff
@@ -31,11 +31,23 @@ collections:
...
+```
+
+#### Install RHSSO from the Customer Portal, when the following variables are defined
```
Comment (sabre1041) [inline README.md]:
```suggestion
#### Install RHSSO from the Red Hat Customer Portal, when the following variables are defined
```

File: roles/keycloak/README.md [hunk 5]
```diff
@@ -66,20 +66,45 @@ The following variables are _required_ only when `keycloak_db_enabled` is True:
 |`keycloak_db_user` | username for connecting to postgres | `keycloak-user` |
 |`keycloak_db_pass` | password for connecting to postgres | `keycloak-pass` |
 
+The following variables can be used to install Keycloak or Red Hat Single Sign-On from local path:
+| Variable | Description | Default |
+|:---------|:------------|:---------|
+|`rhsso_zip_file_local_path` | Full local path of Red Hat Single Sign-On zip file  | `tmp/rhsso/rh-sso-7.5-server-dist.zip` |
+|`keycloak_zip_file_local_path` | Full local path of Keycloak zip file  | `/tmp/keycloak/keycloak-16.1.0.zip` |
```
Comment (sabre1041) [inline roles/keycloak/README.md]: Why do we have separate paths for RHSSO/keycloak. Why not just have a path for local file?

File: roles/keycloak/defaults/main.yml [hunk 6]
```diff
@@ -6,7 +6,7 @@ keycloak_download_url: https://downloads.jboss.org/keycloak/{{ keycloak_version
 keycloak_installdir: "{{ keycloak_dest }}/keycloak-{{ keycloak_version }}"
 
 ### Configuration specific to Red Hat Single Sing-On
-keycloak_rhsso_enable: "{{ True if rhsso_rhn_id is defined else False }}"
+keycloak_rhsso_enable: "{{ True if rhsso_rhn_id is defined or rhsso_zip_file_local_path is defined  or rhsso_source_download_url is defined else False }}"
```
Comment (sabre1041) [inline roles/keycloak/defaults/main.yml]: Why cant this be used as the determination of whether this is upstream or keycloak?

**Review by sabre1041 — CHANGES_REQUESTED** [review]
While reviewing the collection, I found myself taking a step back and determining the matrix of possibilities: These include distribution types and download methods:

1. First determine what distribution type you want to install (upstream keycloak/RHSSO)
2. How are you going obtain the source (RHN/remote location/local on control node)

These factors drive the logic of the collection

File: roles/keycloak/defaults/main.yml (hunk 6 above)
Comment (codecontribu) [reply roles/keycloak/defaults/main.yml]: Actually that's how we are determining if it is upstream or RHSSO

**Comment by codecontribu** [comment]
Addressed all comments

**Pull request merged**